### DISCUSSION TITLE
High cpu usage on Windows Server 2022 Core

### DISCUSSION THREAD

**Discussion by marukka** [bluenviron/mediamtx#2296] [Questions]
Is it expected that Media MTX use 25-50% of a Intel Xeon E-2134 CPU core to proxy a single RTSP stream? This seems a bit excessive even when it is proxying it to no one. I am aware of sourceOnDemand, I'm just using proxying to no one as an example. For comparison ffmpeg on the same computer uses 0-3% of a CPU core to copy the RTSP stream from Media MTX (or the camera directly) to a .MKV file. 

If it matters, the RTSP streams are from Reolink E1 Outdoor PoE cameras. They use a H.265 CODEC, 3840x2160 resolution, 25 FPS and a AAC audio stream. According to VLC the content bitrate bounces around from about 5mbit/sec to 8mbit/sec. Media MTX 1.0.3 is running on a Windows Server 2022 Core VM. The config file is pretty much stock, other than disabling other protocols, setting the RTSP port and path/source.

Config file - [mediamtx.txt](https://github.com/bluenviron/mediamtx/files/12518287/mediamtx.txt)

**Comment by aler9** [comment]
Hello, MediaMTX consumes a negligible CPU amount when idle or when receiving streams, not certainly in the order of 25-50%. In order to check whether the server is consuming CPU, you can use pprof as described in the README:

https://github.com/bluenviron/mediamtx#pprof

Otherwise open an issue and provide instructions on how to replicate the problem.

**Comment by marukka** [comment]
@aler9 here is that pprof output you requested
![Task Manager/console output](https://github.com/bluenviron/mediamtx/assets/3067075/f5d51fc5-6306-4a59-85a2-d0bbc9024f52)
[pprof files.zip](https://github.com/bluenviron/mediamtx/files/12566492/pprof.files.zip)

**Comment by marukka** [comment]
So I tried spinning up a VMware Photon Linux 4 VM and ran your MediaMTX binary on it. CPU utilization is 1-3% with the exact same config as my previous screenshot other than pprof being disabled.
![2023-09-09 14_25_15-OpenSSH SSH client](https://github.com/bluenviron/mediamtx/assets/3067075/ee29d86b-76b4-46e3-88b6-de73e622aa9b)

**Comment by aler9** [comment]
The pprof output you attached shows that MediaMTX consumes 0.47% of available CPU even on Windows:

```
File: C:\MediaMTX\mediamtx.exe
Build ID: C:\MediaMTX\mediamtx.exe2023-09-04 21:14:40.2681364 -0500 CDT
Type: cpu
Time: Sep 9, 2023 at 6:55pm (UTC)
Duration: 30.01s, Total samples = 140ms ( 0.47%)
Showing nodes accounting for 140ms, 100% of 140ms total
      flat  flat%   sum%        cum   cum%
      40ms 28.57% 28.57%       40ms 28.57%  runtime.cgocall
      10ms  7.14% 35.71%       20ms 14.29%  github.com/bluenviron/mediamtx/internal/core.(*rtspSource).run.func6.1.1
      10ms  7.14% 42.86%       10ms  7.14%  net.wrapSyscallError (inline)
      10ms  7.14% 50.00%       10ms  7.14%  runtime.duffcopy
      10ms  7.14% 57.14%       10ms  7.14%  runtime.ifaceeq
      10ms  7.14% 64.29%       20ms 14.29%  runtime.mallocgc
      10ms  7.14% 71.43%       30ms 21.43%  runtime.netpoll
      10ms  7.14% 78.57%       10ms  7.14%  runtime.scanobject
      10ms  7.14% 85.71%       10ms  7.14%  runtime.stdcall5
      10ms  7.14% 92.86%       10ms  7.14%  runtime.stdcall6
      10ms  7.14%   100%       10ms  7.14%  runtime.unlock2
         0     0%   100%       10ms  7.14%  bufio.(*Reader).Peek
         0     0%   100%       50ms 35.71%  bufio.(*Reader).Read
         0     0%   100%       10ms  7.14%  bufio.(*Reader).fill
         0     0%   100%       20ms 14.29%  github.com/bluenviron/gortsplib/v4.(*clientFormat).readRTPTCP
         0     0%   100%       30ms 21.43%  github.com/bluenviron/gortsplib/v4.(*clientMedia).readRTPTCPPlay
         0     0%   100%      100ms 71.43%  github.com/bluenviron/gortsplib/v4.(*clientReader).run
         0     0%   100%      100ms 71.43%  github.com/bluenviron/gortsplib/v4.(*clientReader).runInner
         0     0%   100%       60ms 42.86%  github.com/bluenviron/gortsplib/v4/pkg/base.(*InterleavedFrame).Unmarshal
         0     0%   100%       60ms 42.86%  github.com/bluenviron/gortsplib/v4/pkg/bytecounter.(*ByteCounter).Read
         0     0%   100%       70ms 50.00%  github.com/bluenviron/gortsplib/v4/pkg/conn.(*Conn).Read
         0     0%   100%       60ms 42.86%  github.com/bluenviron/gortsplib/v4/pkg/conn.(*Conn).ReadInterleavedFrame (inline)
         0     0%   100%       10ms  7.14%  github.com/bluenviron/mediamtx/internal/stream.(*Stream).WriteRTPPacket
         0     0%   100%       10ms  7.14%  github.com/bluenviron/mediamtx/internal/stream.(*streamFormat).writeRTPPacket
         0     0%   100%       10ms  7.14%  github.com/bluenviron/mediamtx/internal/stream.(*streamFormat).writeUnit
         0     0%   100%       10ms  7.14%  github.com/bluenviron/mediamtx/internal/stream.unitSize
         0     0%   100%       50ms 35.71%  internal/poll.(*FD).Read
         0     0%   100%       40ms 28.57%  internal/poll.(*FD).Read.func1
         0     0%   100%       50ms 35.71%  internal/poll.execIO
         0     0%   100%       50ms 35.71%  io.ReadAtLeast
         0     0%   100%       50ms 35.71%  io.ReadFull (inline)
         0     0%   100%       60ms 42.86%  net.(*conn).Read
         0     0%   100%       60ms 42.86%  net.(*netFD).Read
         0     0%   100%       10ms  7.14%  runtime.deductAssistCredit
         0     0%   100%       40ms 28.57%  runtime.findRunnable
         0     0%   100%       10ms  7.14%  runtime.gcAssistAlloc
         0     0%   100%       10ms  7.14%  runtime.gcAssistAlloc.func1
         0     0%   100%       10ms  7.14%  runtime.gcAssistAlloc1
         0     0%   100%       10ms  7.14%  runtime.gcDrainN
         0     0%   100%       10ms  7.14%  runtime.makeslice
         0     0%   100%       40ms 28.57%  runtime.mcall
         0     0%   100%       10ms  7.14%  runtime.newobject
         0     0%   100%       40ms 28.57%  runtime.park_m
         0     0%   100%       40ms 28.57%  runtime.schedule
         0     0%   100%       10ms  7.14%  runtime.systemstack
         0     0%   100%       10ms  7.14%  runtime.unlock (inline)
         0     0%   100%       10ms  7.14%  runtime.unlockWithRank (inline)
         0     0%   100%       40ms 28.57%  syscall.Syscall9
         0     0%   100%       40ms 28.57%  syscall.SyscallN
         0     0%   100%       40ms 28.57%  syscall.WSARecv

```

**Comment by aler9** [comment]
my guess: is it the antivirus or Window Defender running?

**Comment by marukka** [comment]
I think whatever API call pprof is using is having problems, perhaps it is only profiling your code rather than also including whatever external APIs you're using. This is reproducible on the VM was well with Task Manager. The "Performance" tab's graphs will show low CPU utilization, whereas the "Details" tab and Performance Monitor (perfmon.msc) will both show high CPU utilization. See the attached video for an example. Are you able to get a copy of the free VMware Player (or paid VMware Workstation or ESXi) in your environment? I could upload an OVA of the VM it is running on so you can easily reproduce this issue yourself.

![2023-09-09 18_03_05-Performance Monitor](https://github.com/bluenviron/mediamtx/assets/3067075/c5ea6b01-88f1-4543-b556-bf9e35f6b1e3)

https://github.com/bluenviron/mediamtx/assets/3067075/05116990-27a2-4b0e-9d01-4c8ba8faa549

**Comment by marukka** [comment]
High CPU utilization can also be confirmed by using Visual Studio's debugging functions. I installed VS Remote Debugger on this VM and was able to profile the process. With 6 RTSP streams added, CPU utilization is using almost the entire core at times with no one connected to MediaMTX to view the streams.
![2023-09-09 18_26_36-Videos](https://github.com/bluenviron/mediamtx/assets/3067075/4d50471d-b8ac-4cee-8994-a9d4fecf4d69)

**Comment by marukka** [comment]
And here is another screenshot, this time with a single camera and more importantly with the external symbols loaded. Its been a while since I've used VisualStudio but if i'm reading this right, it looks like it is a problem with your net code or more likely whatever API you're using:
![2023-09-09 18_40_43-CPU Usage (Debugging) - Microsoft Visual Studio LTSC 17 6 (Administrator)](https://github.com/bluenviron/mediamtx/assets/3067075/51308606-c087-45d1-b242-998f3276f85c)

[a comment by an automated account is not shown]
